runtime/src/qio/qio.c

```diff
@@ -2879,38 +2886,113 @@ qioerr _qio_buffered_write(qio_channel_t* ch, const void* ptr, ssize_t len, ssiz
   //if( _right_mark_start(ch) >= ch->end_pos ) return QIO_EEOF;
   if (qio_channel_offset_unlocked(ch) >= ch->end_pos) return QIO_EEOF;
 
-  while ((remaining > 0) && !eof) {
-    if ((ch->bufIoMax > 0) && (remaining > ch->bufIoMax)) {
-      toWrite = ch->bufIoMax;
-    } else {
-      toWrite = remaining;
+  // if possible make a direct system call instead of buffering
+  if ( len > qio_write_unbuffered_threshold &&                        // the write is large enough
+       method != QIO_METHOD_MMAP && method != QIO_METHOD_MEMORY &&    // we aren't using mmap or mem
+       ( ch->mark_cur == 0 || ch->mark_cur == 1 ) &&                  // buffering isn't waiting for a commit/revert
+       ch->chan_info == NULL                                          // there is no IO plugin
+  ) {
+    // printf("Unbuffered Write ------------------------\n");
+
+    // check if this write will exceed EOF
+    if ( ch->end_pos < INT64_MAX && _right_mark_start(ch) + len > ch->end_pos ) {
+      eof = 1;
+      err = QIO_EEOF;
+      goto error;
     }
 
-    // make sure we have buffer space. (require calls advance_cached)
-    err = _qio_channel_require_unlocked(ch, toWrite, 1);
-    eof = 0;
-    if( qio_err_to_int(err) == EEOF ) eof = 1;
-    else if( err ) goto error;
-
-    // figure out the end of the data to copy
-    start = _right_mark_start_iter(ch);
-    end = start;
-    gotlen = qbuffer_iter_num_bytes_after(&ch->buf, end);
-    if( toWrite < gotlen ) gotlen = toWrite;
-    qbuffer_iter_advance(&ch->buf, &end, gotlen);
-
-    // now copy the data in to the buffer.
-    err = qbuffer_copyin(&ch->buf, start, end, (char *) ptr + (len-remaining),
-                         gotlen);
+    // flush the write-behind portion of the buffer before attempting to write more
+    err = _qio_channel_flush_qio_unlocked(ch);
     if( err ) goto error;
 
-    // now move start forward.
-    _set_right_mark_start(ch, end.offset);
+    // trim any excess bytes from the current buffer part
```

Comment:
Is this required for correctness? I think this is telling it to drop the whole buffer?

Reply:
IIRC, I was getting some weird zero padding in the output without this. 

Based on the comment in `qio_channel_seek` where I grabbed this code, I initially figured it would be okay to use it here since I just called `_qio_buffered_behind` (inside `_qio_channel_flush_qio_unlocked`):

https://github.com/chapel-lang/chapel/blob/288a7905b70e5447e0982083d4f8c89a95f0dd48/runtime/src/qio/qio.c#L3973-L3980

But I did not realize the whole buffer would be discarded. I'll try to remove this and see if I can get rid of that padding issue a different way.

Reply:
I think it could be OK to leave this in, but that might present a performance problem. The issue is that if we keep the buffers allocated, we have to sortof renumber where they start. But, the buffer code should be able to do that.